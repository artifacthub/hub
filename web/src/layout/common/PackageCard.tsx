import isUndefined from 'lodash/isUndefined';
import React from 'react';
import { Link } from 'react-router-dom';

import { Package, SearchFiltersURL } from '../../types';
import buildPackageURL from '../../utils/buildPackageURL';
import styles from './PackageCard.module.css';
import PackageInfo from './PackageInfo';

interface Props {
  package: Package;
  saveScrollPosition?: () => void;
  searchUrlReferer?: SearchFiltersURL;
  fromStarredPage?: boolean;
  visibleSignedBadge?: boolean;
}

const PackageCard = (props: Props) => (
  <div className="col-12 py-sm-3 py-2" role="listitem">
    <div className={`card h-100 ${styles.card}`}>
      <Link
        data-testid="link"
        className={`text-decoration-none text-reset ${styles.link}`}
        onClick={() => {
          if (!isUndefined(props.saveScrollPosition)) {
            props.saveScrollPosition();
          }
        }}
        to={{
          pathname: buildPackageURL(props.package),
          state: { searchUrlReferer: props.searchUrlReferer, fromStarredPage: props.fromStarredPage },
        }}
      >
        <div className={`card-body d-flex flex-column ${styles.body}`}>
          <PackageInfo
            package={props.package}
            visibleSignedBadge={props.visibleSignedBadge}
            withPackageLinks={false}
            breakpointForInfoSection="lg"
          />
        </div>
      </Link>
    </div>
  </div>
);

export default PackageCard;
