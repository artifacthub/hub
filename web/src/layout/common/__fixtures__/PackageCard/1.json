{
  "packageId": "id",
  "name": "test",
  "normalizedName": "test",
  "displayName": "Pretty name",
  "description": "desc",
  "logoImageId": "imageId",
  "appVersion": "1.0.0",
  "securityReportSummary": {
    "low": 0,
    "high": 7,
    "medium": 1,
    "unknown": 0,
    "critical": 1
  },
  "repository": {
    "repositoryId": "0acb228c-17ab-4e50-85e9-ffc7102ea423",
    "kind": 0,
    "name": "stable",
    "url": "repoUrl",
    "userAlias": "user"
  }
}
