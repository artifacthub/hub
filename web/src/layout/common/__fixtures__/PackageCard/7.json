{
  "packageId": "id",
  "kind": 0,
  "name": "test",
  "normalizedName": "test",
  "displayName": "Pretty name",
  "description": "desc",
  "logoImageId": "imageId",
  "appVersion": "1.0.0",
  "userAlias": "usertest",
  "chartRepository": {
    "name": "stable",
    "displayName": null,
    "url": "repoUrl"
  }
}
