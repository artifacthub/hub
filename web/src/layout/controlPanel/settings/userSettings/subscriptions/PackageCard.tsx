import isNull from 'lodash/isNull';
import isUndefined from 'lodash/isUndefined';
import React from 'react';
import { Link } from 'react-router-dom';

import { EventKind, Package } from '../../../../../types';
import buildPackageURL from '../../../../../utils/buildPackageURL';
import { SubscriptionItem, SUBSCRIPTIONS_LIST } from '../../../../../utils/data';
import Image from '../../../../common/Image';
import RepositoryIcon from '../../../../common/RepositoryIcon';
import styles from './PackageCard.module.css';

interface Props {
  package: Package;
  changeSubscription: (packageId: string, kind: EventKind, isActive: boolean, packageName: string) => void;
}

const PackageCard = (props: Props) => {
  return (
    <div className="py-2" role="listitem">
      <div className={`card h-100 ${styles.card}`}>
        <div>
          <div className={`card-body position-relative ${styles.body}`}>
            <Link
              data-testid="packageCardLink"
              className={`text-decoration-none ${styles.link}`}
              to={{
                pathname: buildPackageURL(props.package),
              }}
            >
              <div className="d-flex align-items-start justify-content-between mb-3">
                <div className={`d-flex align-items-center flex-grow-1 ${styles.truncateWrapper}`}>
                  <div
                    className={`d-flex align-items-center justify-content-center overflow-hidden p-1 ${styles.imageWrapper}`}
                  >
                    <Image
                      imageId={props.package.logoImageId}
                      alt={`Logo ${props.package.displayName || props.package.name}`}
                      className={styles.image}
                    />
                  </div>

                  <div className={`ml-3 flex-grow-1 ${styles.truncateWrapper}`}>
                    <div className={`card-title font-weight-bolder mb-2 ${styles.title}`}>
                      <div className="h5 d-flex flex-row align-items-center">
                        {props.package.displayName || props.package.name}
                      </div>
                    </div>

                    <div className={`card-subtitle d-flex flex-wrap mw-100 mt-1 ${styles.subtitle}`}>
                      {!isUndefined(props.package.repository.organizationName) &&
                        props.package.repository.organizationName && (
                          <div className="mr-2 text-truncate">
                            <span className="text-muted text-uppercase mr-1">Org:</span>
                            <p className="d-inline mb-0 text-dark">
                              {props.package.repository.organizationDisplayName ||
                                props.package.repository.organizationName}
                            </p>
                          </div>
                        )}

                      {!isNull(props.package.repository.userAlias) && (
                        <div className="mr-2 text-truncate">
                          <span className="text-muted text-uppercase mr-1">User:</span>
                          <p className="d-inline mb-0 text-dark">{props.package.repository.userAlias}</p>
                        </div>
                      )}

                      <div className="mr-2 text-truncate">
                        <span className="text-muted text-uppercase mr-1">Repo:</span>
                        <p className="d-inline mb-0 text-dark">
                          {props.package.repository.displayName || props.package.repository.name}
                        </p>
                      </div>
                    </div>
                  </div>
                </div>

                <div className={`d-flex align-items-center text-uppercase ${styles.kind}`}>
                  <RepositoryIcon className={styles.icon} kind={props.package.repository.kind} />
                </div>
              </div>
            </Link>

            {SUBSCRIPTIONS_LIST.map((subs: SubscriptionItem) => {
              const isActive = !isUndefined(props.package.eventKinds) && props.package.eventKinds.includes(subs.kind);

              return (
                <div
                  className={`d-flex flex-row align-items-center ${styles.wrapper}`}
                  key={`card_${props.package.normalizedName}_${subs.kind}`}
                >
                  <div className={`custom-control custom-switch ${styles.clickable}`}>
                    <input
                      data-testid={`${subs.name}MobileBtn`}
                      id={subs.name}
                      type="checkbox"
                      className={`custom-control-input ${styles.checkbox}`}
                      disabled={!subs.enabled}
                      onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                        e.preventDefault();
                        e.stopPropagation();

                        props.changeSubscription(
                          props.package.packageId,
                          subs.kind,
                          isActive,
                          props.package.displayName || props.package.name
                        );
                      }}
                      checked={isActive}
                    />
                    <label data-testid={`${subs.name}_label`} className="custom-control-label" htmlFor={subs.name}>
                      <div className="d-flex align-items-center">
                        {subs.icon} <span className="ml-1">{subs.title}</span>
                      </div>
                    </label>
                  </div>
                </div>
              );
            })}
          </div>
        </div>
      </div>
    </div>
  );
};

export default PackageCard;
