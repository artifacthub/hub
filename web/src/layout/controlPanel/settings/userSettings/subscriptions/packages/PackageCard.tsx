import isUndefined from 'lodash/isUndefined';
import React from 'react';
import { Link } from 'react-router-dom';

import { EventKind, Package } from '../../../../../../types';
import buildPackageURL from '../../../../../../utils/buildPackageURL';
import { PACKAGE_SUBSCRIPTIONS_LIST, SubscriptionItem } from '../../../../../../utils/data';
import Image from '../../../../../common/Image';
import RepositoryIcon from '../../../../../common/RepositoryIcon';
import styles from './PackageCard.module.css';

interface Props {
  package: Package;
  changeSubscription: (packageId: string, kind: EventKind, isActive: boolean, packageName: string) => void;
}

const PackageCard = (props: Props) => {
  return (
    <div className="py-2" role="listitem">
      <div className={`card cardWithHover h-100 ${styles.card}`}>
        <div>
          <div className={`card-body position-relative ${styles.body}`}>
            <Link
              data-testid="packageCardLink"
              className={`text-decoration-none text-reset ${styles.link}`}
              to={{
                pathname: buildPackageURL(
                  props.package.normalizedName,
                  props.package.repository,
                  props.package.version!
                ),
              }}
            >
              <div className="d-flex align-items-start justify-content-between mb-3">
                <div className={`d-flex align-items-center flex-grow-1 ${styles.truncateWrapper}`}>
                  <div
                    className={`d-flex align-items-center justify-content-center overflow-hidden p-1 ${styles.imageWrapper} imageWrapper`}
                  >
                    <Image
                      imageId={props.package.logoImageId}
                      alt={`Logo ${props.package.displayName || props.package.name}`}
                      className={styles.image}
                      kind={props.package.repository.kind}
                    />
                  </div>

                  <div className={`ml-3 flex-grow-1 ${styles.truncateWrapper}`}>
                    <div className={`card-title font-weight-bolder mb-2 ${styles.title}`}>
                      <div className="h5 d-flex flex-row align-items-center">
                        {props.package.displayName || props.package.name}
                      </div>
                    </div>

                    <div className={`card-subtitle d-flex flex-wrap mw-100 mt-1 ${styles.subtitle}`}>
                      {props.package.repository.organizationName && (
                        <div className="mr-2 text-truncate">
                          <span className="text-muted text-uppercase mr-1">Org:</span>
                          <p className="d-inline mb-0 text-dark">
                            {props.package.repository.organizationDisplayName ||
                              props.package.repository.organizationName}
                          </p>
                        </div>
                      )}

                      {props.package.repository.userAlias && (
                        <div className="mr-2 text-truncate">
                          <span className="text-muted text-uppercase mr-1">User:</span>
                          <p className="d-inline mb-0 text-dark">{props.package.repository.userAlias}</p>
                        </div>
                      )}

                      <div className="mr-2 text-truncate">
                        <span className="text-muted text-uppercase mr-1">Repo:</span>
                        <p className="d-inline mb-0 text-dark">
                          {props.package.repository.displayName || props.package.repository.name}
                        </p>
                      </div>
                    </div>
                  </div>
                </div>

                <div className={`d-flex align-items-center text-uppercase ${styles.kind}`}>
                  <RepositoryIcon className={styles.icon} kind={props.package.repository.kind} />
                </div>
              </div>
            </Link>

            {PACKAGE_SUBSCRIPTIONS_LIST.map((subs: SubscriptionItem) => {
              const isActive = !isUndefined(props.package.eventKinds) && props.package.eventKinds.includes(subs.kind);

              return (
                <div
                  className="d-flex flex-row align-items-center"
                  key={`card_${props.package.normalizedName}_${subs.kind}`}
                >
                  <div className={`custom-control custom-switch ${styles.clickable}`}>
                    <input
                      data-testid={`${subs.name}MobileBtn`}
                      id={`${props.package.name}_${subs.name}`}
                      type="checkbox"
                      className={`custom-control-input ${styles.checkbox}`}
                      disabled={!subs.enabled}
                      onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                        e.preventDefault();
                        e.stopPropagation();

                        props.changeSubscription(
                          props.package.packageId,
                          subs.kind,
                          isActive,
                          props.package.displayName || props.package.name
                        );
                      }}
                      checked={isActive}
                    />
                    <label
                      data-testid={`${subs.name}_label`}
                      className="custom-control-label"
                      htmlFor={`${props.package.name}_${subs.name}`}
                    >
                      <div className="d-flex align-items-center">
                        {subs.icon} <span className="ml-1">{subs.title}</span>
                      </div>
                    </label>
                  </div>
                </div>
              );
            })}
          </div>
        </div>
      </div>
    </div>
  );
};

export default PackageCard;
