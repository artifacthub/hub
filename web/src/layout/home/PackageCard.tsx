import React from 'react';
import { Link } from 'react-router-dom';

import { Package } from '../../types';
import buildPackageURL from '../../utils/buildPackageURL';
import PackageInfo from '../common/PackageInfo';
import styles from './PackageCard.module.css';

interface Props {
  package: Package;
  className?: string;
}

const PackageCard = (props: Props) => (
  <div className={`col-12 col-xxl-5 py-sm-3 py-2 px-0 px-xxl-3 position-relative ${props.className}`}>
    <div className={`card cardWithHover h-100 ${styles.card}`}>
      <Link
        data-testid="link"
        className={`text-decoration-none text-reset h-100 ${styles.link}`}
        to={{
          pathname: buildPackageURL(props.package.normalizedName, props.package.repository, props.package.version!),
        }}
      >
        <div className={`card-body d-flex flex-column h-100 ${styles.body}`}>
          <PackageInfo package={props.package} withPackageLinks={false} />
        </div>
      </Link>
    </div>
  </div>
);

export default PackageCard;
