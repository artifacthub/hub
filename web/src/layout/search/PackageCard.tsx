import isNull from 'lodash/isNull';
import React from 'react';
import { Link, useHistory } from 'react-router-dom';

import { Package, PackageKind, SearchFiltersURL } from '../../types';
import buildPackageURL from '../../utils/buildPackageURL';
import prepareQueryString from '../../utils/prepareQueryString';
import Image from '../common/Image';
import PackageIcon from '../common/PackageIcon';
import styles from './PackageCard.module.css';

interface Props {
  package: Package;
  saveScrollPosition: () => void;
  searchUrlReferer: SearchFiltersURL | null;
}

const PackageCard = (props: Props) => {
  const history = useHistory();
  return (
    <div className="col-12 py-sm-3 py-2" role="listitem">
      <div className={`card h-100 ${styles.card}`}>
        <Link
          data-testid="link"
          className={`text-decoration-none ${styles.link}`}
          onClick={props.saveScrollPosition}
          to={{
            pathname: buildPackageURL(props.package),
            state: props.searchUrlReferer,
          }}
        >
          <div className={`card-body ${styles.body}`}>
            <div className="d-flex align-items-start justify-content-between mb-3">
              <div className={`d-flex align-items-center flex-grow-1 ${styles.truncateWrapper}`}>
                <div
                  className={`d-flex align-items-center justify-content-center overflow-hidden p-1 ${styles.imageWrapper}`}
                >
                  <Image
                    imageId={props.package.logoImageId}
                    alt={`Logo ${props.package.displayName || props.package.name}`}
                    className={styles.image}
                  />
                </div>

                <div className={`ml-3 flex-grow-1 ${styles.truncateWrapper}`}>
                  <div className={`card-title font-weight-bolder mb-2 ${styles.title}`}>
                    <div className="h5 d-flex flex-row align-items-center">
                      {props.package.displayName || props.package.name}
                      {!isNull(props.package.deprecated) && props.package.deprecated && (
                        <div
                          className={`d-none d-sm-flex badge badge-pill ml-2 mt-1 text-uppercase ${styles.deprecatedBadge}`}
                        >
                          Deprecated
                        </div>
                      )}
                    </div>
                  </div>

                  <div className={`card-subtitle d-flex flex-wrap mw-100 mt-1 ${styles.subtitle}`}>
                    {(() => {
                      switch (props.package.kind) {
                        case PackageKind.Chart:
                          return (
                            <>
                              <div className="mr-2 text-truncate">
                                <span className="text-muted text-uppercase mr-1">Repository: </span>
                                <button
                                  data-testid="repoLink"
                                  className={`p-0 border-0 ${styles.link}`}
                                  onClick={(e) => {
                                    e.preventDefault();
                                    history.push({
                                      pathname: '/search',
                                      search: prepareQueryString({
                                        pageNumber: 1,
                                        filters: {
                                          repo: [props.package.chartRepository!.name],
                                        },
                                        deprecated: isNull(props.package.deprecated) ? false : props.package.deprecated,
                                      }),
                                      state: { fromSearchCard: true },
                                    });
                                  }}
                                >
                                  <u>
                                    {props.package.chartRepository!.displayName || props.package.chartRepository!.name}
                                  </u>
                                </button>
                              </div>

                              <div className="mr-2 text-truncate">
                                <span className="text-muted text-uppercase mr-1">Version: </span>
                                {props.package.version || '-'}
                              </div>

                              <div className="text-truncate">
                                <span className="text-muted text-uppercase mr-1">App version: </span>
                                {props.package.appVersion || '-'}
                              </div>
                            </>
                          );

                        case PackageKind.Falco:
                        case PackageKind.Opa:
                          return (
                            <div className="text-truncate">
                              <span className="text-muted text-uppercase mr-1">Version: </span>
                              {props.package.version || '-'}
                            </div>
                          );

                        default:
                          return null;
                      }
                    })()}
                  </div>
                </div>
              </div>

              <div className={`d-flex align-items-center text-uppercase ${styles.kind}`}>
                <PackageIcon className={styles.icon} kind={props.package.kind} />
              </div>
            </div>

            <p className={`mb-0 card-text overflow-hidden ${styles.description}`}>{props.package.description}</p>

            {!isNull(props.package.deprecated) && props.package.deprecated && (
              <div className={`d-inline d-sm-none badge badge-pill mt-1 text-uppercase ${styles.deprecatedBadge}`}>
                Deprecated
              </div>
            )}
          </div>
        </Link>
      </div>
    </div>
  );
};

export default PackageCard;
