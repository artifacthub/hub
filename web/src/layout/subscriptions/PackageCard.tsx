import isNull from 'lodash/isNull';
import isUndefined from 'lodash/isUndefined';
import React from 'react';
import { FaCheck } from 'react-icons/fa';
import { Link, useHistory } from 'react-router-dom';

import { NotificationKind, Package, PackageKind } from '../../types';
import buildPackageURL from '../../utils/buildPackageURL';
import { SubscriptionItem, SUBSCRIPTIONS_LIST } from '../../utils/data';
import prepareQueryString from '../../utils/prepareQueryString';
import Image from '../common/Image';
import OrganizationInfo from '../common/OrganizationInfo';
import PackageIcon from '../common/PackageIcon';
import styles from './PackageCard.module.css';

interface Props {
  package: Package;
  changeSubscription: (packageId: string, kind: NotificationKind, isActive: boolean, packageName: string) => void;
}

const PackageCard = (props: Props) => {
  const history = useHistory();
  return (
    <div className="col-12 py-sm-3 py-2" role="listitem">
      <div className={`card h-100 ${styles.card}`}>
        <Link
          data-testid="packageLink"
          className={`text-decoration-none ${styles.link}`}
          to={{
            pathname: buildPackageURL(props.package),
          }}
        >
          <div className={`card-body position-relative ${styles.body}`}>
            <div className="d-flex align-items-start justify-content-between mb-3">
              <div className={`d-flex align-items-center flex-grow-1 ${styles.truncateWrapper}`}>
                <div
                  className={`d-flex align-items-center justify-content-center overflow-hidden p-1 ${styles.imageWrapper}`}
                >
                  <Image
                    imageId={props.package.logoImageId}
                    alt={`Logo ${props.package.displayName || props.package.name}`}
                    className={styles.image}
                  />
                </div>

                <div className={`ml-3 flex-grow-1 ${styles.truncateWrapper}`}>
                  <div className={`card-title font-weight-bolder mb-2 ${styles.title}`}>
                    <div className="h5 d-flex flex-row align-items-center">
                      {props.package.displayName || props.package.name}
                    </div>
                  </div>

                  <div className={`card-subtitle d-flex flex-wrap mw-100 mt-1 ${styles.subtitle}`}>
                    {!isUndefined(props.package.organizationName) && props.package.organizationName && (
                      <OrganizationInfo
                        className="mr-2"
                        organizationName={props.package.organizationName}
                        organizationDisplayName={props.package.organizationDisplayName}
                        deprecated={props.package.deprecated}
                        visibleLegend
                      />
                    )}

                    {!isNull(props.package.userAlias) && (
                      <div className="mr-2 text-truncate">
                        <span className="text-muted text-uppercase mr-1">User:</span>
                        <button
                          data-testid="userLink"
                          className={`p-0 border-0 ${styles.link}`}
                          onClick={(e) => {
                            e.preventDefault();
                            history.push({
                              pathname: '/packages/search',
                              search: prepareQueryString({
                                pageNumber: 1,
                                filters: {
                                  user: [props.package.userAlias!],
                                },
                                deprecated: false,
                              }),
                            });
                          }}
                        >
                          <u>{props.package.userAlias}</u>
                        </button>
                      </div>
                    )}

                    {(() => {
                      switch (props.package.kind) {
                        case PackageKind.Chart:
                          return (
                            <>
                              <div className="mr-2 text-truncate">
                                <span className="text-muted text-uppercase mr-1">Repo:</span>
                                <button
                                  data-testid="repoLink"
                                  className={`p-0 border-0 ${styles.link}`}
                                  onClick={(e) => {
                                    e.preventDefault();
                                    history.push({
                                      pathname: '/packages/search',
                                      search: prepareQueryString({
                                        pageNumber: 1,
                                        filters: {
                                          repo: [props.package.chartRepository!.name],
                                        },
                                        deprecated: false,
                                      }),
                                    });
                                  }}
                                >
                                  <u>
                                    {props.package.chartRepository!.displayName || props.package.chartRepository!.name}
                                  </u>
                                </button>
                              </div>
                            </>
                          );

                        default:
                          return null;
                      }
                    })()}
                  </div>
                </div>
              </div>

              <div className={`d-flex align-items-center text-uppercase ${styles.kind}`}>
                <PackageIcon className={styles.icon} kind={props.package.kind} />
              </div>
            </div>

            {SUBSCRIPTIONS_LIST.map((subs: SubscriptionItem) => {
              const isActive =
                !isUndefined(props.package.notificationKinds) && props.package.notificationKinds.includes(subs.kind);

              return (
                <button
                  data-testid={`${subs.name}MobileBtn`}
                  key={`card_${props.package.normalizedName}_${subs.kind}`}
                  className="btn btn-primary btn-sm"
                  disabled={!subs.enabled}
                  onClick={(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
                    e.preventDefault();
                    e.stopPropagation();

                    props.changeSubscription(
                      props.package.packageId,
                      subs.kind,
                      isActive,
                      props.package.displayName || props.package.name
                    );
                  }}
                >
                  <div className="d-flex flex-row align-items-center">
                    {subs.icon} <span className="ml-1">{subs.title}</span>
                    <span className={`m-0 p-0 ml-3 badge badge-pill badge-light ${styles.badge}`}>
                      <small className="text-success">{isActive && <FaCheck />}</small>
                    </span>
                  </div>
                </button>
              );
            })}
          </div>
        </Link>
      </div>
    </div>
  );
};

export default PackageCard;
