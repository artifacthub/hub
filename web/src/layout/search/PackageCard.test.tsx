import React from 'react';
import { render, queryByText, fireEvent } from '@testing-library/react';
import PackageCard from './PackageCard';
import { BrowserRouter as Router } from 'react-router-dom';
import prepareQuerystring from '../../utils/prepareQueryString';
import { Package } from '../../types';
import buildPackageURL from '../../utils/buildPackageURL';

const getMockPackage = (fixtureId: string): Package => {
  return require(`./__fixtures__/packageCard/${fixtureId}.json`) as Package;
};

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const mockSaveScrollPosition = jest.fn();

const defaultProps = {
  saveScrollPosition: mockSaveScrollPosition,
  searchUrlReferer: null,
}

describe('PackageCard', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders correctly', () => {
    const mockPackage = getMockPackage('1');
    const { asFragment } = render(
      <Router>
        <PackageCard {...defaultProps} package={mockPackage} />
      </Router>
    );
    expect(asFragment).toMatchSnapshot();
  });

  describe('Image', () => {
    it('renders package logo', () => {
      const mockPackage = getMockPackage('2');
      const { queryByAltText } = render(
        <Router>
          <PackageCard {...defaultProps} package={mockPackage} />
        </Router>
      );
      const image = queryByAltText(`Logo ${mockPackage.displayName}`);
      expect(image).toBeInTheDocument();
    });

    it('renders placeholder when imageId is null', () => {
      const mockPackage = getMockPackage('3');

      const { queryByAltText } = render(
        <Router>
          <PackageCard {...defaultProps} package={mockPackage} />
        </Router>
      );
      const image = queryByAltText(`Logo ${mockPackage.displayName}`);
      expect(image).toBeInTheDocument();
      expect((image as HTMLImageElement).src).toBe('http://localhost/kubernetes_grey.svg');
    });
  });

  describe('Title', () => {
    it('renders display name', () => {
      const mockPackage = getMockPackage('4');

      const { queryByText } = render(
        <Router>
          <PackageCard {...defaultProps} package={mockPackage} />
        </Router>
      );
      const title = queryByText(mockPackage.displayName!);
      expect(title).toBeInTheDocument();
    });

    it('renders name when display name is null', () => {
      const mockPackage = getMockPackage('5');

      const { queryByText } = render(
        <Router>
          <PackageCard {...defaultProps} package={mockPackage} />
        </Router>
      );
      const title = queryByText(mockPackage.name!);
      expect(title).toBeInTheDocument();
    });
  });

  describe('Chart repository button', () => {
    it('renders repository link', () => {
      const mockPackage = getMockPackage('6');

      const { queryByTestId, queryByAltText } = render(
        <Router>
          <PackageCard {...defaultProps} package={mockPackage} />
        </Router>
      );
      const button = queryByTestId('repoLink');
      expect(button).toBeInTheDocument();
      const icon = queryByAltText('Icon');
      expect(icon).toBeInTheDocument();
      expect((icon as HTMLImageElement).src).toBe('http://localhost/helm.svg');
      fireEvent.click(button!);
      expect(mockHistoryPush).toHaveBeenCalledTimes(1);
      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/search',
        search: prepareQuerystring({
          pageNumber: 1,
          filters: {
            repo: [mockPackage.chartRepository!.name]
          },
          deprecated: false,
        }),
        state: { fromSearchCard: true },
      });
    });

    it('does not render repository link when chart kind is not Helm Chart', () => {
      const mockPackage = getMockPackage('7');

      const { queryByRole } = render(
        <Router>
          <PackageCard {...defaultProps} package={mockPackage} />
        </Router>
      );
      const button = queryByRole('button');
      expect(button).toBeNull();
    });
  });

  describe('Detail', () => {
    it('opens detail page', () => {
      const mockPackage = getMockPackage('8');
      const urlReferer = {
        text: 'test',
        filters: {},
        pageNumber: 1,
        deprecated: false,
      };
      const { queryByTestId } = render(
        <Router>
          <PackageCard
            {...defaultProps}
            package={mockPackage}
            searchUrlReferer={urlReferer}
          />
        </Router>
      );
      const link = queryByTestId('link');
      expect(link).toBeInTheDocument();
      fireEvent.click(link!);
      expect(mockSaveScrollPosition).toHaveBeenCalledTimes(1);
      expect(location.pathname).toBe(buildPackageURL(mockPackage));
    });
  });
});
