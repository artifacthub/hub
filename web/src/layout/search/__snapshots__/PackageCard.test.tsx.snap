// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageCard renders correctly 1`] = `[Function]`;
